File: core/internal/leet/flexlayout.go
package leet

import (
	"math"
	"strings"

	"charm.land/lipgloss/v2"
)

// stackSectionID identifies a vertically stacked pane in the main content area.
type stackSectionID int

const (
	stackSectionMetrics stackSectionID = iota
	stackSectionSystemMetrics
	stackSectionMedia
	stackSectionConsoleLogs
	stackSectionCount
)

// stackSectionSpec describes one pane in a vertical stack.
type stackSectionSpec struct {
	ID      stackSectionID
	Visible bool
	Height  int
	Flex    bool
}

// stackSectionLayout stores the computed origin and height of one pane.
type stackSectionLayout struct {
	Y      int
	Height int
}

// verticalStackLayout is a small flexbox-style layout for the central column.
//
// Fixed-height panes (system/media/logs) keep their current animated heights.
// The optional flex pane (metrics) consumes the remaining height after gaps.
type verticalStackLayout struct {
	TotalHeight  int
	VisibleCount int
	Sections     [stackSectionCount]stackSectionLayout
}

// computeVerticalStackLayout computes a top-to-bottom stack with a 1-line gap
// between adjacent visible panes.
func computeVerticalStackLayout(
	totalHeight int,
	specs ...stackSectionSpec,
) verticalStackLayout {
	layout := verticalStackLayout{TotalHeight: max(totalHeight, 0)}

	visible := make([]stackSectionSpec, 0, len(specs))
	fixedHeight := 0
	flexIndex := -1
	for _, spec := range specs {
		if !spec.Visible {
			continue
		}
		spec.Height = max(spec.Height, 0)
		if spec.Flex {
			flexIndex = len(visible)
		} else {
			fixedHeight += spec.Height
		}
		visible = append(visible, spec)
	}

	layout.VisibleCount = len(visible)
	if len(visible) == 0 {
		return layout
	}

	gapLines := max(len(visible)-1, 0)
	remaining := max(layout.TotalHeight-fixedHeight-gapLines, 0)
	if flexIndex >= 0 {
		visible[flexIndex].Height = remaining
	}

	y := 0
	for i, spec := range visible {
		layout.Sections[spec.ID] = stackSectionLayout{Y: y, Height: spec.Height}
		y += spec.Height
		if i < len(visible)-1 {
			y++
		}
	}

	return layout
}

func (l verticalStackLayout) Height(id stackSectionID) int {
	if id < 0 || id >= stackSectionCount {
		return 0
	}
	return l.Sections[id].Height
}

func (l verticalStackLayout) Y(id stackSectionID) int {
	if id < 0 || id >= stackSectionCount {
		return 0
	}
	return l.Sections[id].Y
}

const (
	// sidebarDragMinWidth is the narrowest a sidebar override can make it.
	// Deliberately smaller than SidebarMinWidth so users can shrink
	// sidebars below the default clamp.
	sidebarDragMinWidth = 20

	// mainDragMinWidth is the narrowest the main content column can be
	// squeezed to by a sidebar override.
	mainDragMinWidth = 24
)

// expandedSidebarWidth returns a sidebar's expanded width: a user-dragged
// fraction of the terminal width when set (non-zero), or the golden-ratio
// default. Dragged widths may go below the default minimum but always leave
// room for the main content column.
func expandedSidebarWidth(terminalWidth int, oppositeVisible bool, frac float64) int {
	if frac > 0 {
		maxW := max(terminalWidth-mainDragMinWidth, sidebarDragMinWidth)
		w := int(math.Round(float64(terminalWidth) * frac))
		return clamp(w, sidebarDragMinWidth, maxW)
	}

	ratio := SidebarWidthRatio
	if oppositeVisible {
		ratio = SidebarWidthRatioBoth
	}
	return clamp(int(float64(terminalWidth)*ratio), SidebarMinWidth, SidebarMaxWidth)
}

// minFlexMetricsHeight is the shortest the flex metrics section can be
// squeezed to by the fixed panes below it.
const minFlexMetricsHeight = 5

// fitSidebarFractions clamps sidebar override fractions so both expanded
// sidebars together leave the main content column its minimum width. Only
// overridden (non-zero) fractions are adjusted; the golden-ratio defaults
// already fit jointly.
func fitSidebarFractions(
	width int,
	leftVisible, rightVisible bool,
	leftFrac, rightFrac float64,
) (float64, float64) {
	if width <= 0 || !leftVisible || !rightVisible ||
		(leftFrac <= 0 && rightFrac <= 0) {
		return leftFrac, rightFrac
	}

	leftW := expandedSidebarWidth(width, true, leftFrac)
	rightW := expandedSidebarWidth(width, true, rightFrac)
	total := width - mainDragMinWidth
	if leftW+rightW <= total {
		return leftFrac, rightFrac
	}

	newLeft, newRight := leftW, rightW
	switch {
	case rightFrac <= 0: // Only the left override can shrink.
		newLeft = total - rightW
	case leftFrac <= 0: // Only the right override can shrink.
		newRight = total - leftW
	default: // Shrink both, proportionally.
		newLeft = leftW * total / (leftW + rightW)
		newRight = total - newLeft
	}

	if leftFrac > 0 {
		leftFrac = float64(max(newLeft, sidebarDragMinWidth)) / float64(width)
	}
	if rightFrac > 0 {
		rightFrac = float64(max(newRight, sidebarDragMinWidth)) / float64(width)
	}
	return leftFrac, rightFrac
}

// paneHeightFor returns a stacked pane's expanded height for the given
// override fraction of the terminal height, or the fallback default when no
// override is set. The pane's own SetExpandedHeight enforces its minimum.
func paneHeightFor(frac float64, terminalHeight, fallback int) int {
	if frac <= 0 {
		return fallback
	}
	return int(math.Round(float64(terminalHeight) * frac))
}

// fitStackHeights lowers the fixed stacked panes' heights to fit budget so
// overridden panes cannot crowd out the flex metrics section or overflow
// the stack. Each pane gives up slack above its own minimum, proportionally
// — the panes re-inflate anything below their minimums, which would
// silently undo a plain proportional scale. When even the minimums exceed
// the budget the panes keep their minimums (tiny-terminal degradation the
// renderer crops).
func fitStackHeights(heights, minimums []int, budget int) {
	sum, slack := 0, 0
	for i := range heights {
		sum += heights[i]
		slack += max(heights[i]-minimums[i], 0)
	}
	over := sum - max(budget, 0)
	if over <= 0 {
		return
	}
	if over >= slack {
		for i := range heights {
			heights[i] = min(heights[i], minimums[i])
		}
		return
	}

	left := over
	for i := range heights {
		cut := over * max(heights[i]-minimums[i], 0) / slack
		heights[i] -= cut
		left -= cut
	}
	// Integer-division dust: take the remainder from panes with slack.
	for i := range heights {
		if left == 0 {
			break
		}
		cut := min(left, max(heights[i]-minimums[i], 0))
		heights[i] -= cut
		left -= cut
	}
}

// sidebarContentWidth returns the width available for text content inside a
// sidebar after subtracting the vertical border and both padding columns.
func sidebarContentWidth(totalWidth int) int {
	return max(totalWidth-SidebarOverhead, 0)
}

// sidebarInnerWidth returns the width to pass to the sidebar's lipgloss Style
// (includes content + padding, excludes the border column).
func sidebarInnerWidth(totalWidth int) int {
	return max(totalWidth-SidebarBorderCols, 0)
}

func filterNonEmptySections(sections []string) []string {
	filtered := make([]string, 0, len(sections))
	for _, section := range sections {
		if section == "" {
			continue
		}
		filtered = append(filtered, section)
	}
	return filtered
}

// placeMainColumn pads or crops content to exactly width x height. Mouse
// hit-testing maps screen rows to sections via computeVerticalStackLayout,
// so rendered content must never spill past its reserved rows —
// lipgloss.Place pads short content but never crops tall content (e.g. the
// metrics grid's minimum chart height in a short section).
func placeMainColumn(width, height int, content string) string {
	if width <= 0 || height <= 0 {
		return ""
	}
	if lines := strings.Split(content, "\n"); len(lines) > height {
		content = strings.Join(lines[:height], "\n")
	}
	return lipgloss.Place(width, height, lipgloss.Left, lipgloss.Top, content)
}

File: leet-rs/src/flexlayout.rs
//! A small flexbox-style vertical stack layout for the central column.

use crate::theme::{
    SIDEBAR_MAX_WIDTH, SIDEBAR_MIN_WIDTH, SIDEBAR_OVERHEAD, SIDEBAR_WIDTH_RATIO,
    SIDEBAR_WIDTH_RATIO_BOTH,
};

/// Identifies a vertically stacked pane in the main content area.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum StackSection {
    Metrics = 0,
    SystemMetrics = 1,
    Media = 2,
    ConsoleLogs = 3,
}

pub const STACK_SECTION_COUNT: usize = 4;

/// Describes one pane in a vertical stack.
#[derive(Debug, Clone, Copy)]
pub struct StackSectionSpec {
    pub id: StackSection,
    pub visible: bool,
    pub height: i32,
    pub flex: bool,
}

/// The computed origin and height of one pane.
#[derive(Debug, Clone, Copy, Default)]
pub struct StackSectionLayout {
    pub y: i32,
    pub height: i32,
}

/// A computed top-to-bottom stack.
///
/// Fixed-height panes (system/media/logs) keep their current animated
/// heights. The optional flex pane (metrics) consumes the remaining height
/// after gaps.
#[derive(Debug, Clone, Copy, Default)]
pub struct VerticalStackLayout {
    pub total_height: i32,
    pub visible_count: usize,
    sections: [StackSectionLayout; STACK_SECTION_COUNT],
}

impl VerticalStackLayout {
    pub fn height(&self, id: StackSection) -> i32 {
        self.sections[id as usize].height
    }

    pub fn y(&self, id: StackSection) -> i32 {
        self.sections[id as usize].y
    }
}

/// Computes a top-to-bottom stack with a 1-line gap between adjacent visible
/// panes.
pub fn compute_vertical_stack_layout(
    total_height: i32,
    specs: &[StackSectionSpec],
) -> VerticalStackLayout {
    let mut layout = VerticalStackLayout {
        total_height: total_height.max(0),
        ..Default::default()
    };

    let mut visible: Vec<StackSectionSpec> = Vec::with_capacity(specs.len());
    let mut fixed_height = 0;
    let mut flex_index: Option<usize> = None;
    for spec in specs {
        if !spec.visible {
            continue;
        }
        let mut spec = *spec;
        spec.height = spec.height.max(0);
        if spec.flex {
            flex_index = Some(visible.len());
        } else {
            fixed_height += spec.height;
        }
        visible.push(spec);
    }

    layout.visible_count = visible.len();
    if visible.is_empty() {
        return layout;
    }

    let gap_lines = (visible.len() as i32 - 1).max(0);
    let remaining = (layout.total_height - fixed_height - gap_lines).max(0);
    if let Some(i) = flex_index {
        visible[i].height = remaining;
    }

    let mut y = 0;
    let last = visible.len() - 1;
    for (i, spec) in visible.iter().enumerate() {
        layout.sections[spec.id as usize] = StackSectionLayout {
            y,
            height: spec.height,
        };
        y += spec.height;
        if i < last {
            y += 1;
        }
    }

    layout
}

/// The full width of an expanded sidebar for the given terminal width.
pub fn expanded_sidebar_width(terminal_width: i32, opposite_visible: bool) -> i32 {
    let ratio = if opposite_visible {
        SIDEBAR_WIDTH_RATIO_BOTH
    } else {
        SIDEBAR_WIDTH_RATIO
    };
    ((terminal_width as f64 * ratio) as i32).clamp(SIDEBAR_MIN_WIDTH, SIDEBAR_MAX_WIDTH)
}

/// Minimum sidebar width while resizing with the mouse. Narrower than the
/// default clamp so users can deliberately shrink a sidebar.
pub const SIDEBAR_DRAG_MIN_WIDTH: i32 = 20;

/// Minimum central column width preserved while resizing sidebars.
pub const MAIN_DRAG_MIN_WIDTH: i32 = 24;

/// The expanded sidebar width, honoring a user-set fraction of the terminal
/// width when present.
pub fn sidebar_width_for(
    terminal_width: i32,
    opposite_visible: bool,
    fraction: Option<f64>,
) -> i32 {
    match fraction {
        Some(f) => {
            let max = (terminal_width - MAIN_DRAG_MIN_WIDTH).max(SIDEBAR_DRAG_MIN_WIDTH);
            ((terminal_width as f64 * f).round() as i32).clamp(SIDEBAR_DRAG_MIN_WIDTH, max)
        }
        None => expanded_sidebar_width(terminal_width, opposite_visible),
    }
}

/// The width available for text content inside a sidebar after subtracting
/// the vertical border and both padding columns.
pub fn sidebar_content_width(total_width: i32) -> i32 {
    (total_width - SIDEBAR_OVERHEAD).max(0)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn stack_layout_with_flex() {
        let layout = compute_vertical_stack_layout(
            40,
            &[
                StackSectionSpec {
                    id: StackSection::Metrics,
                    visible: true,
                    height: 0,
                    flex: true,
                },
                StackSectionSpec {
                    id: StackSection::SystemMetrics,
                    visible: true,
                    height: 10,
                    flex: false,
                },
                StackSectionSpec {
                    id: StackSection::ConsoleLogs,
                    visible: false,
                    height: 5,
                    flex: false,
                },
            ],
        );
        assert_eq!(layout.visible_count, 2);
        // Metrics gets 40 - 10 fixed - 1 gap = 29.
        assert_eq!(layout.height(StackSection::Metrics), 29);
        assert_eq!(layout.y(StackSection::Metrics), 0);
        assert_eq!(layout.y(StackSection::SystemMetrics), 30);
    }
}
